src/server/pfs/server/driver.go

```diff
@@ -3310,28 +3285,78 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 	return req, nil
 }
 
-func forEachPutFile(server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) error {
+func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (oneOff bool, repo string, branch string, err error) {
 	limiter := limit.New(client.DefaultMaxConcurrentStreams)
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
 	var req *pfs.PutFileRequest
-	var err error
 	var eg errgroup.Group
+	var rawCommitID string
+	var commitID string
+
 	for req, err = server.Recv(); err == nil; req, err = server.Recv() {
 		req := req
 		if req.File != nil {
+			// For the first file, dereference the commit ID if needed. For
+			// subsequent files, ensure that they're all referencing the same
+			// commit ID, and replace their commit objects to all refer to the
+			// same thing.
+			if rawCommitID == "" {
+				commit := req.File.Commit
+				repo = commit.Repo.Name
+				// The non-dereferenced commit ID. Used to ensure that all
+				// subsequent requests use the same value.
+				rawCommitID = commit.ID
+				// oneOff is true if we're creating the commit as part of this
+				// put-file
+				oneOff = false
+				// inspectCommit will replace file.Commit.ID with an actual
+				// commit ID if it's a branch. So we want to save it first.
+				if !uuid.IsUUIDWithoutDashes(commit.ID) {
+					branch = commit.ID
+				}
+				// inspect the commit where we're adding files and figure out
+				// if this is a one-off put-file. 
+				// - if 'commit' refers to an open commit                -> not oneOff
+				// - otherwise (i.e. branch with closed HEAD or no HEAD) -> yes oneOff
+				// Note that if commit is a specific commit ID, it must be
+				// open for this call to succeed
+				commitInfo, err := d.inspectCommit(pachClient, commit, pfs.CommitState_STARTED)
+				if err != nil {
+					if (!isNotFoundErr(err) && !isNoHeadErr(err)) || branch == "" {
+						return false, "", "", err
+					}
+					oneOff = true
+				}
+				if commitInfo != nil && commitInfo.Finished != nil {
+					if branch == "" {
+						return false, "", "", pfsserver.ErrCommitFinished{commit}
+					}
+					oneOff = true
+				}
+				commitID = commit.ID
+			} else if req.File.Commit.ID != rawCommitID {
+				err = fmt.Errorf("All requests in a put files call must have the same commit ID; expected '%s', got '%s'", rawCommitID, req.File.Commit.ID)
+				return false, "", "", err
+			} else if req.File.Commit.Repo.Name != repo {
+				err = fmt.Errorf("All requests in a put files call must have the same repo name; expected '%s', got '%s'", repo, req.File.Commit.Repo.Name)
+				return false, "", "", err
+			} else {
+				req.File.Commit.ID = commitID
+			}
+
```

Comment:
In the old implementation, we were peeking the first put file request and dereferencing its commit ID, but not doing so for subsequent requests. As a result, on an open commit to a branch, the first file in a put file request would go through (because the commit ID was dereferenced), but not subsequent requests.

With this change, the first put file request's commit ID is still dereferenced, but we now update subsequent requests to use the dereferenced commit ID.